### ISSUE TITLE
Add `active_workflows` to the project update schema

### ISSUE THREAD

**Issue by simoneduca** [zooniverse/Panoptes#2679]
## Current behaviour
When trying to update a link with endpoints like https://panoptes.zooniverse.org/api/projects/376/links/active_workflows a 404 error is returned.

I think the `active_workflows` link should be added to the list below to fix this problem.
https://github.com/zooniverse/Panoptes/blob/8efd6fb0fbf5e89daf2b890431bc7bcab5a126d7/app/schemas/project_update_schema.rb#L124

**Comment by srallen** [comment]
I disagree. Being able to directly update those links shouldn't be necessary. The API serializes the active workflows linked to the project based on the workflows `active: bool` state, so it should be the workflow that is the PUT request is made to for updating its active state and then the API handles updating the serialized links for us.

**Comment by camallen** [comment]
I think allowing modification of the `active_workflows` link has merit. Almost all other resource links work this way and it's surprising that some don't, we should limit surprise for the API consumer. 

Another bigger use case is that changing the workflow state impacts another resource, specifically the project resource `active_workflows` links, however the response object of the workflows PUT operation is a workflow resource not the updated project. Hence the client will have an outdated local copy of the project resource and can't rely on that state being correct. To know the new state of the project, the client would then have to get an updated copy from the server at the expense of another network request (and remembering to do this in code). Instead of these client issues, we can allow the links relationship to modify the resource we are interested in (project in this case). This request (if successful) would return the updated project resource state in the response body and the client will be in sync with the server. 

In summary, if the client needs to understand that changes on one resource (workflow) changes state on another (project) then we should be allowing that change through the linking resource (project) in this case. We should still allow the state of the workflow resource update as well and then it is up to the client to decide which resource they need to update for their use case.

**Comment by simoneduca** [comment]
What made me lose quite some time yesterday was definitely the surprise effect. I would find this change helpful.

**Comment by srallen** [comment]
Please correct me if I'm wrong, however, the suggestion is that instead of making a network request to get the updated project, we would manually update the project links? I personally would prefer the former than the latter. I think there maybe greater risk of state being inconsistent with each other if the client has to update both the workflow and the project link when making the workflow active or inactive.

**Comment by camallen** [comment]
It really depends on the resource you are working on, if you are working on a project and want to modify it's `active_workflows` state then we should allow this, likewise for workflows.

Most our link relations work like this in panoptes, e.g. linking a set to a workflow, a subject to a set, etc. Concurrent updates should be handled by the API and you don't have to supply the whole relation state, rather they will append API side. 

>  I think there maybe greater risk of state being inconsistent with each other if the client has to update both the workflow and the project link when making the workflow active or inactive.

I can't speak for the client code but I do think it depends on context, if the state that is being tested is the belongs to the project resource use the links to update the state, if it's the workflow then use the active attribute.

**Comment by srallen** [comment]
I understand having consistent behavior with links, but this is a link that also implies a certain state on the workflow. None of the other link relationship as far as I know do this. If we allow the client to update this link and only this link is updated, does this mean the active state on the workflow changes? Or are we going to have the client entirely manage adding/removing the link at the time the workflow activeness is toggled? We need a callback, either a GET request or a request to update the links of the project, to update the parent component state.

**Comment by camallen** [comment]
There are examples of API calls that modify other resources but not in a way that requires immediate reflection of that state, e.g. linking subjects to sets creates a whole new resource (set member subjects) as well as modifies counts on workflows & projects. Another is retiring subjects for a workflow changes state across workflows & projects as well as subject workflow status. 

>If we allow the client to update this link and only this link is updated, does this mean the active state on the workflow changes?

API side - yes. If you want to then fetch the updated workflow that's a reasonable approach.

> We need a callback, either a GET request or a request to update the links of the project, to update the parent component state.

Yep - and this points to what i think is the crux of the matter, the API is currently sharding this state across two resources. Upon reflection, the active state **coupling** between workflow & project could have been better modeled, instead of being on the workflow it should have been an explicit internal join model that has a single point of access via the project resource. This way you would only have to update / get the project resource to manage the state. This better reflects that actual relationship of active workflows for a project.

I can look at doing this, especially if it will simplify client code. A quick fix would be to use a callback to get the state on the other resource that the client needs.

**Comment by marten** [comment]
>>If we allow the client to update this link and only this link is updated, does this mean the active state on the workflow changes?
>API side - yes. If you want to then fetch the updated workflow that's a reasonable approach.

But that leaves you with the exact same problem as you described: the state of a workflow resource implictly changes by linking it, and the client doesn't know that it needs to invalidate it and reload.

**Comment by camallen** [comment]
Sure - I think i've outlined the issues with this approach as it stands for coupled state, I've also tried out outline a solution too.

**Comment by srallen** [comment]
I don't feel too strongly about this, but fwiw, I think it's more clear to me that it should be the workflow resource that is modified and then we uncache and get the project resource rather than updating the project and its links since its the state of the workflow that is changing.
